zio-http/src/main/scala/zio/http/codec/internal/EncoderDecoder.scala

```diff
@@ -243,8 +318,39 @@ private[codec] object EncoderDecoder                   {
       } else if (jsonDecoders.length == 1) {
         jsonDecoders(0)(body).map { result => inputs(0) = result }
       } else {
-        ZIO.foreachDiscard(jsonDecoders.zipWithIndex) { case (decoder, index) =>
-          decoder(body).map { result => inputs(index) = result }
+        body.asMultipartFormStream.flatMap { form =>
+          form.fields.runForeach { field =>
+            indexByName.get(field.name) match {
+              case Some(idx) =>
+                flattened.content(idx) match {
+                  case BodyCodec.Multiple(schema, _, _) if schema == Schema[Byte] =>
+                    field match {
+                      case FormField.Binary(_, data, _, _, _)          =>
+                        inputs(idx) = ZStream.fromChunk(data)
+                      case FormField.StreamingBinary(_, _, _, _, data) =>
+                        inputs(idx) = data
+                      case FormField.Text(_, value, _, _)              =>
+                        inputs(idx) = ZStream.fromChunk(Chunk.fromArray(value.getBytes(Charsets.Utf8)))
+                      case FormField.Simple(_, value)                  =>
+                        inputs(idx) = ZStream.fromChunk(Chunk.fromArray(value.getBytes(Charsets.Utf8)))
+                    }
+                    ZIO.unit
+                  case _                                                          =>
+                    formFieldDecoders(idx)(field).map { result => inputs(idx) = result }
+                }
+              case None      =>
+                ZIO.fail(HttpCodecError.MalformedBody(s"Unexpected multipart/form-data field: ${field.name}"))
+            }
+          }.zipRight {
+            ZIO.attempt {
+              for (idx <- inputs.indices) {
+                if (inputs(idx) == null)
+                  throw HttpCodecError.MalformedBody(
+                    s"Missing multipart/form-data field (${flattened.content(idx).name.getOrElse("field" + idx.toString)}",
```

Comment:
This is a translation from `idx` to field name. Could be useful to precompute that, unless it's only used in the error case here.